File: Examples/ObjectiveC/CustomAnnotationModelExample.m
//
//  CustomAnnotationModelExample.m
//  Examples
//
//  Created by Jason Wray on 5/20/16.
//  Copyright © 2016 Mapbox. All rights reserved.
//

#import "CustomAnnotationModelExample.h"
#import "CustomAnnotationModels.h"
@import Mapbox;

NSString *const MBXExampleCustomAnnotationModel = @"CustomAnnotationModelExample";

@interface CustomAnnotationModelExample () <MGLMapViewDelegate>

@end

@implementation CustomAnnotationModelExample

- (void)viewDidLoad {
    [super viewDidLoad];

    MGLMapView *mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds];
    mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    mapView.styleURL = [MGLStyle lightStyleURLWithVersion:9];
    mapView.tintColor = [UIColor darkGrayColor];
    mapView.zoomLevel = 1;
    mapView.delegate = self;
    [self.view addSubview:mapView];

    // Point Annotation
    CustomPointAnnotation *point = [[CustomPointAnnotation alloc] init];
    point.coordinate = CLLocationCoordinate2DMake(0, 0);
    point.title = @"Custom Point Annotation";
    // Set the custom `image` and `reuseIdentifier` properties, later used in the `mapView:imageForAnnotation:` delegate method.
    point.image = [self dotWithSize:20];
    point.reuseIdentifier = @"yeOldeGreyDot";

    // Polyline
    // Create a coordinates array with all of the coordinates for our polyline.
    CLLocationCoordinate2D coordinates[] = {
        CLLocationCoordinate2DMake(35, -25),
        CLLocationCoordinate2DMake(20, -30),
        CLLocationCoordinate2DMake( 0, -25),
        CLLocationCoordinate2DMake(-15,  0),
        CLLocationCoordinate2DMake(-45, 10),
        CLLocationCoordinate2DMake(-45, 40),
    };
    NSUInteger numberOfCoordinates = sizeof(coordinates) / sizeof(CLLocationCoordinate2D);

    CustomPolyline *polyline = [CustomPolyline polylineWithCoordinates:coordinates count:numberOfCoordinates];//
    // Set the custom `color` property, later used in the `mapView:strokeColorForShapeAnnotation:` delegate method.
    polyline.color = [UIColor darkGrayColor];

    // Add both annotations to the map.
    [mapView addAnnotations:@[point, polyline]];
}

- (UIImage *)dotWithSize:(CGFloat)size {
    CGRect rect = CGRectMake(0, 0, size, size);
    CGFloat strokeWidth = 1;

    UIGraphicsBeginImageContextWithOptions(rect.size, NO, [[UIScreen mainScreen] scale]);

    UIBezierPath *ovalPath = [UIBezierPath bezierPathWithOvalInRect:CGRectInset(rect, strokeWidth, strokeWidth)];
    [UIColor.darkGrayColor setFill];
    [ovalPath fill];

    [UIColor.whiteColor setStroke];
    ovalPath.lineWidth = strokeWidth;
    [ovalPath stroke];

    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return image;
}

#pragma mark - MGLMapViewDelegate methods

-(MGLAnnotationImage *)mapView:(MGLMapView *)mapView imageForAnnotation:(id<MGLAnnotation>)annotation {
    if ([annotation isKindOfClass:[CustomPointAnnotation class]]) {
        CustomPointAnnotation *point = (CustomPointAnnotation *)annotation;
        MGLAnnotationImage *annotationImage = [mapView dequeueReusableAnnotationImageWithIdentifier:point.reuseIdentifier];

        if (annotationImage) {
            // The annotatation image has already been cached, just reuse it.
            return annotationImage;
        } else if (point.image && point.reuseIdentifier) {
            // Create a new annotation image.
            return [MGLAnnotationImage annotationImageWithImage:point.image reuseIdentifier:point.reuseIdentifier];
        }
    }

    // Fallback to the default marker image.
    return nil;
}

- (UIColor *)mapView:(MGLMapView *)mapView strokeColorForShapeAnnotation:(MGLShape *)annotation {
    if ([annotation isKindOfClass:[CustomPolyline class]]) {
        // Return orange if the polyline does not have a custom color.
        return [(CustomPolyline *)annotation color] ?: [UIColor orangeColor];
    }

    // Fallback to the default tint color.
    return mapView.tintColor;
}

-(BOOL)mapView:(MGLMapView *)mapView annotationCanShowCallout:(id<MGLAnnotation>)annotation {
    return YES;
}

@end

File: Examples/Swift/CustomAnnotationModelExample.swift
//
//  CustomAnnotationModelExample.swift
//  Examples
//
//  Created by Jason Wray on 5/20/16.
//  Copyright © 2016 Mapbox. All rights reserved.
//

import Mapbox

@objc(CustomAnnotationModelExample_Swift)

class CustomAnnotationModelExample_Swift: UIViewController, MGLMapViewDelegate {
    override func viewDidLoad() {
        super.viewDidLoad()

        let mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = [.FlexibleWidth, .FlexibleHeight]
        mapView.styleURL = MGLStyle.lightStyleURLWithVersion(9)
        mapView.tintColor = .darkGrayColor()
        mapView.zoomLevel = 1
        mapView.delegate = self
        view.addSubview(mapView)

        // Polyline
        // Create a coordinates array with all of the coordinates for our polyline.
        var coordinates = [
            CLLocationCoordinate2DMake(35, -25),
            CLLocationCoordinate2DMake(20, -30),
            CLLocationCoordinate2DMake( 0, -25),
            CLLocationCoordinate2DMake(-15,  0),
            CLLocationCoordinate2DMake(-45, 10),
            CLLocationCoordinate2DMake(-45, 40),
        ]

        let polyline = CustomPolyline(coordinates: &coordinates, count: UInt(coordinates.count))
        // Set the custom `color` property, later used in the `mapView:strokeColorForShapeAnnotation:` delegate method.
        polyline.color = .darkGrayColor()

        // Add the polyline to the map. Note that this method name is singular.
        mapView.addAnnotation(polyline)

        // Point Annotations
        // Add a custom point annotation for every coordinate (vertex) in the polyline.
        var pointAnnotations = [CustomPointAnnotation]()
        for coordinate in coordinates {
            let count = pointAnnotations.count + 1
            let point = CustomPointAnnotation(coordinate: coordinate,
                                              title: "Custom Point Annotation \(count)",
                                              subtitle: nil)

            // Set the custom `image` and `reuseIdentifier` properties, later used in the `mapView:imageForAnnotation:` delegate method.
            // Create a unique reuse identifier for each new annotation image.
            point.reuseIdentifier = "customAnnotation\(count)"
            // This dot image grows in size as more annotations are added to the array.
            point.image = dot(size:5 * count)

            // Append each annotation to the array, which will be added to the map all at once.
            pointAnnotations.append(point)
        }

        // Add the point annotations to the map. This time the method name is plural.
        // If you have multiple annotations to add, batching their addition to the map is more efficient.
        mapView.addAnnotations(pointAnnotations)
    }

    func dot(size size: Int) -> UIImage {
        let floatSize = CGFloat(size)
        let rect = CGRectMake(0, 0, floatSize, floatSize)
        let strokeWidth: CGFloat = 1

        UIGraphicsBeginImageContextWithOptions(rect.size, false, UIScreen.mainScreen().scale)

        let ovalPath = UIBezierPath(ovalInRect: CGRectInset(rect, strokeWidth, strokeWidth))
        UIColor.darkGrayColor().setFill()
        ovalPath.fill()

        UIColor.whiteColor().setStroke()
        ovalPath.lineWidth = strokeWidth
        ovalPath.stroke()

        let image: UIImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return image
    }

    // MARK: - MGLMapViewDelegate methods

    func mapView(mapView: MGLMapView, imageForAnnotation annotation: MGLAnnotation) -> MGLAnnotationImage? {
        if let point = annotation as? CustomPointAnnotation,
            image = point.image,
            reuseIdentifier = point.reuseIdentifier {

            if let annotationImage = mapView.dequeueReusableAnnotationImageWithIdentifier(reuseIdentifier) {
                // The annotatation image has already been cached, just reuse it.
                return annotationImage
            } else {
                // Create a new annotation image.
                return MGLAnnotationImage(image: image, reuseIdentifier: reuseIdentifier)
            }
        }

        // Fallback to the default marker image.
        return nil
    }

    func mapView(mapView: MGLMapView, strokeColorForShapeAnnotation annotation: MGLShape) -> UIColor {
        if let annotation = annotation as? CustomPolyline {
            // Return orange if the polyline does not have a custom color.
            return annotation.color ?? .orangeColor()
        }

        // Fallback to the default tint color.
        return mapView.tintColor
    }

    func mapView(mapView: MGLMapView, annotationCanShowCallout annotation: MGLAnnotation) -> Bool {
        return true
    }
}
